{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b78fd064-1dfb-41d5-8300-17130fd2c374",
   "metadata": {},
   "outputs": [],
   "source": [
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "####  Mouse C3KO-RichLi snRNA\n",
    "####  2023-05-22 by Yi Zhao (Texas Heart Institute, US)\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "####  Initiate  ####\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~----\n",
    "Ver <- '0'\n",
    "Step <- 'PART21_Milo_Enrichment'\n",
    "Project <- '2023_neoc3ko_rli'\n",
    "\n",
    "Code_dir <- paste0('/Volumes/shire/project/', Project, '/code/mouse_v', Ver, '/')\n",
    "\n",
    "source(paste0(Code_dir, 'src/bioinformatics.R'))\n",
    "source(paste0(Code_dir, 'src/scRNAseq.R'))\n",
    "source(paste0(Code_dir, 'src/scATACseq.R'))\n",
    "source(paste0(Code_dir, 'mouse_v', Ver, '.helper_functions.R'))\n",
    "suppressMessages(library('miloR'))\n",
    "\n",
    "InitiateProject('Rivendell', Ver, Step, 'mouse', Project, 'shire')\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cb2b7d6-753b-41ba-9575-d1428ed0f4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "####  Load data  ####\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~----\n",
    "full.srt <- readRDS('integrated/PART11.cardio_cells.srt.rds')\n",
    "full.srt <- DietSeurat(full.srt, assays = 'RNA', scale.data = F, dimreducs = names(full.srt@reductions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "39d9aed2-aba8-452a-8675-ddcda56f9ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "####  MiloR Process  ####\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~----\n",
    "RunMilo <- function(seurat, k = 20, d = 50, alpha = 0.1, reduction){\n",
    "        sce <- as.SingleCellExperiment(seurat)\n",
    "        milo.meta <- seurat@meta.data\n",
    "        milo.obj <- Milo(sce)\n",
    "        reducedDim(milo.obj, \"UMAP\") <- reducedDim(sce, str_to_upper(reduction)) ## because Milo only uses UMAP\n",
    "        milo.obj <- buildGraph(milo.obj, k = k, d = d, reduced.dim = 'SCVI')\n",
    "        milo.obj <- makeNhoods(milo.obj, k = k, d = d, refined = TRUE, prop = 0.2, reduced_dims = 'SCVI')\n",
    "        milo.obj <- calcNhoodDistance(milo.obj, d = d, reduced.dim = 'SCVI')\n",
    "        milo.obj <- countCells(milo.obj, samples = \"group1\", meta.data = milo.meta)\n",
    "\n",
    "        milo.design <- as.data.frame(xtabs(~ group2 + group1, data = milo.meta))\n",
    "        milo.design <- milo.design[milo.design$Freq > 0, ]\n",
    "        milo.design <- distinct(milo.design)\n",
    "        rownames(milo.design) <- milo.design$group1\n",
    "\n",
    "        milo.res <- testNhoods(milo.obj, design = ~ group2, design.df = milo.design, reduced.dim = 'SCVI')\n",
    "        milo.obj <- buildNhoodGraph(milo.obj)\n",
    "        return(list(milo.obj, milo.res))\n",
    "}\n",
    "plotNhoodGraphDA.my <- function (x, milo_res, alpha = 0.05, res_column = \"logFC\", ...)\n",
    "{\n",
    "        if (is.character(layout)) {\n",
    "                if (!layout %in% names(reducedDims(x))) {\n",
    "                        stop(layout, \"is not in readucedDim(x) - choose a different layout\")\n",
    "                }\n",
    "        }\n",
    "        signif_res <- milo_res\n",
    "        signif_res[signif_res$SpatialFDR > alpha, res_column] <- 0\n",
    "        colData(x)[res_column] <- NA\n",
    "        colData(x)[unlist(nhoodIndex(x)[signif_res$Nhood]), res_column] <- signif_res[, res_column]\n",
    "        plotNhoodGraph(x, colour_by = res_column, ...)\n",
    "}\n",
    "####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b9362207-a32f-438f-ac2a-15e3d3a30f1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"c3_p2m\" \"c3_p2s\"\n",
      "[1] \"c3_p2m\" \"wt_p2m\"\n",
      "[1] \"c3_p2m\" \"wt_p2s\"\n",
      "[1] \"c3_p2s\" \"wt_p2m\"\n",
      "[1] \"c3_p2s\" \"wt_p2s\"\n",
      "[1] \"wt_p2m\" \"wt_p2s\"\n"
     ]
    }
   ],
   "source": [
    "data.group <- c('c3_p2m', 'c3_p2s', 'wt_p2m', 'wt_p2s')\n",
    "for (i in 1:4){\n",
    "    for (m in i+1:4){\n",
    "        if(m <= 4){\n",
    "#            print(c(i,m))\n",
    "            print(c(data.group[i],data.group[m]))\n",
    "        }\n",
    "#        print(data.group[i],data.group[m])\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "571e2d46-44f0-45ff-ad2f-358c3454d516",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n",
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n",
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n",
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n",
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n",
      "Constructing kNN graph with k:10\n",
      "\n",
      "Checking valid object\n",
      "\n",
      "Running refined sampling with reduced_dim\n",
      "\n",
      "Checking meta.data validity\n",
      "\n",
      "Counting cells in neighbourhoods\n",
      "\n",
      "Using TMM normalisation\n",
      "\n",
      "Performing spatial FDR correction withk-distance weighting\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "Idents(full.srt) <- 'group2'\n",
    "for (i in 1:4){\n",
    "    for (m in i+1:4){\n",
    "        if(m <= 4){\n",
    "            sub.srt <- full.srt[, full.srt$group2 %in% c(data.group[i], data.group[m])]\n",
    "\n",
    "            cell.embeddings <- sub.srt@reductions$scVI@cell.embeddings\n",
    "            sub.srt@reductions$scVI <- sub.srt@reductions$sub_clean_scVI_umap\n",
    "            sub.srt@reductions$scVI@cell.embeddings <- cell.embeddings[Cells(sub.srt), ]\n",
    "            sub.srt@reductions$scVI@key <- 'SCVI_'\n",
    "\n",
    "            sub.srt$Cell_type <- droplevels(sub.srt$Cell_type)\n",
    "            sub.srt$Cell_state <- droplevels(sub.srt$Cell_state)\n",
    "            sub.srt$group2 <- droplevels(sub.srt$group2)\n",
    "            sub.srt$group1 <- droplevels(sub.srt$group1)\n",
    "            #change levels: Contrl vs Atrium\n",
    "\n",
    "            ####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "            ####  Milo global  ####\n",
    "            ####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~----\n",
    "            sub.milo <- RunMilo(sub.srt, k = 10, reduction = 'sub_clean_scVI_umap')\n",
    "            milo <- sub.milo\n",
    "            milo[[2]]$logFC[milo[[2]]$logFC > 5] <- 5\n",
    "            milo[[2]]$logFC[milo[[2]]$logFC < -5] <- -5\n",
    "            p <- plotNhoodGraphDA.my(milo[[1]], milo_res = milo[[2]], alpha = 0.1, edge_alpha = 0.1, edge_color = 'grey90') +\n",
    "                    scale_fill_distiller(palette = 'RdBu', limits = c(-5, 5)) +\n",
    "                    theme(aspect.ratio = 1)\n",
    "            p\n",
    "\n",
    "            #mini.srt <- sub.srt[, seq(1, ncol(sub.srt), 5)]\n",
    "            #mini.srt$group1 <- droplevels(mini.srt$group1)\n",
    "            #mini.srt$group2 <- droplevels(mini.srt$group2)\n",
    "\n",
    "\n",
    "            #sub.milo <- RunMilo(mini.srt, k = 10, d = 50, alpha = 0.3, reduction = 'sub_clean_scVI_umap')\n",
    "            #milo <- sub.milo\n",
    "            #milo[[2]]$logFC[milo[[2]]$logFC > 5] <- 5\n",
    "            #milo[[2]]$logFC[milo[[2]]$logFC < -5] <- -5\n",
    "            #p <- plotNhoodGraphDA.my(milo[[1]], milo_res = milo[[2]], alpha = 0.1, edge_alpha = 0.1, edge_color = 'grey90') +\n",
    "            #        scale_fill_distiller(palette = 'RdBu', limits = c(-5, 5)) +\n",
    "            #        theme(aspect.ratio = 1)\n",
    "            #p\n",
    "\n",
    "            gc()\n",
    "            PlotPDF(paste0('1.cardio.umap.',data.group[i],'_vs_',data.group[m],'_clean_milo_result'), 16, 12)\n",
    "            print(p)\n",
    "            dev.off()\n",
    "\n",
    "            mtx <- as.matrix(table(sub.srt$Cell_state, sub.srt$group2))\n",
    "            mtx[, 1] <- mtx[, 1]/Table(sub.srt$group2)[1]\n",
    "            mtx[, 2] <- mtx[, 2]/Table(sub.srt$group2)[2]\n",
    "            ratio <- log2(mtx[, 2]/mtx[, 1])\n",
    "            p2 <- ggplot(data.frame(Ratio = ratio, Subtype = rownames(mtx))) +\n",
    "                    geom_bar(aes(x = Ratio, y = Subtype), stat = 'identity') +\n",
    "                    labs(x = paste0(data.group[i],' vs ',data.group[m]), y = 'Cell subtype') +\n",
    "                    theme_classic() +\n",
    "                    theme(aspect.ratio = 1)\n",
    "            #p2\n",
    "\n",
    "            gc()\n",
    "            PlotPDF(paste0('2.cardio.bar.',data.group[i],'_vs_',data.group[m],'_cell_composition'), 4, 5)\n",
    "            print(p2)\n",
    "            dev.off()\n",
    "\n",
    "            #Table(sub.srt$group2)\n",
    "            #Table(sub.srt$group1)\n",
    "\n",
    "            ####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "            saveRDS(sub.milo, paste0('analysis/PART21.Cardiomyocyte_',data.group[i], '_vs_',data.group[m],'_20pct_downsample_milo_obj_result.rds'))\n",
    "            ####~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "        }\n",
    "#        print(data.group[i],data.group[m])\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
